tests/util/bft.py

```diff
@@ -60,11 +65,45 @@ def interesting_configs(
 
     for config in selected_bft_configs:
         assert config['n'] == 3 * config['f'] + 2 * config['c'] + 1, \
-            "Ivariant breached: n = 3f + 2c + 1"
+            "Invariant breached. Expected: n = 3f + 2c + 1"
 
     return selected_bft_configs
 
 
+def with_trio(async_fn):
+    """ Decorator for running a coroutine (async_fn) with trio. """
+    @wraps(async_fn)
+    def trio_wrapper(*args, **kwargs):
+        return trio.run(async_fn, *args, **kwargs)
+
+    return trio_wrapper
+
+
+def with_bft_network(start_replica_cmd, selected_configs=None):
+    """
+    Runs the decorated async function for all selected BFT configs
+    """
+    def decorator(async_fn):
```

Comment:
I'm a bit confused about how the double nesting works here. It appears to me that when `@with_bft_network(start_replica_cmd)` is used as a decorator it should return `decorator`, which is a standard (non-async) function. Then `@with_trio` runs which will wrap `decorator` and return` trio_wrapper` as the test to run. But when `trio_wrapper` runs it will attempt to call `trio.run` on `decorator` which is not async. However, clearly that can't be the case or none of this would work. So can you please help clarify my understanding? of how things get unwrapped etc...

Reply:
Hey Andrew, thanks, great question! 

The `with_bft_network` function acts as a _decorator factory_. It produces a closure called `decorator` which has access to the `@with_bft_network` parameters. This closure is the actual decorator applied to the underlying async function (`decorator` takes `async_fn` as an argument). From here onwards, it's as if we applied `@decorator` to the underlying function - `wrapper` replaces the decorated function, preserving its identity & metadata using `@wraps`.

Thus, the output when applying `@with_bft_network` is an async function which can be further decorated by using `@with_trio`.

Phew :) I hope that clarifies things. Let me know if I'm missing something.

Reply:
PS: here's a much more detailed explanation, if you wish to dig deeper into the subject:
https://gist.github.com/Zearin/2f40b7b9cfc51132851a#passing-arguments-to-the-decorator

Reply:
Wow. That took me a while to wrap my head around. No pun intended! 

That blog post was great. And it does match your explanation.

 I think the part I was missing was that when the function after the `@` has parentheses, it is evaluated before the `@` is applied to the decorated function. This makes sense, as the `@` needs a function object. It has no idea what to do with those parentheses :D